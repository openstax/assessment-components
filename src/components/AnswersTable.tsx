import { defaultAnswerType } from "../constants";
import { Answer as AnswerType, AnswerDisplayType, ID, ExerciseQuestionData } from "src/types";
import { Answer } from "./Answer";
import { Feedback } from "./Feedback";

export interface AnswersTableProps {
  question: ExerciseQuestionData;
  type?: AnswerDisplayType;
  answer_id?: ID;
  correct_answer_id?: ID | null;
  incorrectAnswerId?: ID;
  answerIdOrder?: ID[],
  feedback_html: string;
  correct_answer_feedback_html?: string;
  answered_count?: number;
  show_all_feedback?: boolean;
  tableFeedbackEnabled?: boolean;
  onChangeAnswer: () => void;
  hideAnswers: boolean;
  hasCorrectAnswer?: boolean;
  onChangeAttempt?: () => void;
  choicesEnabled?: boolean;
  onKeyPress?: () => void;
  contentRenderer?: JSX.Element;
  instructions?: JSX.Element;
}

export const AnswersTable = (props: AnswersTableProps) => {
  let idCounter = 0;

  const {
    question, hideAnswers, type = defaultAnswerType, answered_count, choicesEnabled, correct_answer_id,
    incorrectAnswerId, answer_id, feedback_html, correct_answer_feedback_html,
    show_all_feedback = false, tableFeedbackEnabled, hasCorrectAnswer, onChangeAnswer, onKeyPress, answerIdOrder, instructions
  } = props;
  if (hideAnswers) { return null; }

  const { id } = question;

  const feedback: { index: number, html: string, id: string }[] = [];

  const sortedAnswersByIdOrder = (idOrder: ID[]) => {
    const { answers } = question;
    return answers.slice().sort((a, b) => idOrder.indexOf(a.id) - idOrder.indexOf(b.id));
  }

  const questionAnswerProps = {
    qid: id || `auto-${idCounter++}`,
    answerId: answer_id,
    correctAnswerId: correct_answer_id,
    incorrectAnswerId,
    hasCorrectAnswer,
    onChangeAnswer: onChangeAnswer,
    type,
    answered_count,
    disabled: !choicesEnabled,
    show_all_feedback,
    tableFeedbackEnabled,
    onKeyPress
  };

  const answers = answerIdOrder ? sortedAnswersByIdOrder(answerIdOrder) : question.answers;

  const answersHtml = answers.map((answer, i) => {
    const additionalProps: { answer: AnswerType, iter: number, key: string }
      = {
        answer: {
          ...answer,
          question_id: typeof question.id === 'string' ? parseInt(question.id, 10) : question.id
         },
      iter: i,
      key: `${questionAnswerProps.qid}-option-${i}`,
    };
    const answerProps = Object.assign({}, additionalProps, questionAnswerProps);
    let html: string | undefined;
    let feedbackId: string | undefined;

    if (show_all_feedback && answer.feedback_html && tableFeedbackEnabled) {
      html = answer.feedback_html;
    } else if (answer.id === incorrectAnswerId && feedback_html) {
      html = feedback_html;
    } else if (answer.id === correct_answer_id && correct_answer_feedback_html) {
      html = correct_answer_feedback_html;
    }

    if (html) {
      feedbackId = `feedback-${questionAnswerProps.qid}-${i}`
      feedback.push({ index: i, html, id: feedbackId });
    }

    return (
      <Answer feedbackId={feedbackId} {...answerProps} />
    );
  });

  feedback.forEach((item, i) => {
    const spliceIndex = item.index + i + 1;
    answersHtml.splice(spliceIndex, 0, (
      <Feedback id={item.id} key={spliceIndex} contentRenderer={props.contentRenderer}>
        {item.html}
      </Feedback>
    ));
  });

  return (
    <div className="answers-table">
      {instructions}
      {answersHtml}
    </div>
  );
}
