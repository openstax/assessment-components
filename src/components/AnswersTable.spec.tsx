import { AnswersTable, AnswersTableProps } from './AnswersTable';
import renderer from 'react-test-renderer';
import { answerContent } from '../test/fixtures';

describe('AnswersTable', () => {
  let props: AnswersTableProps;

  beforeEach(() => {
    props = {
      questionId: '1',
      answers: [{
        id: '1',
        correctness: undefined,
        content_html: 'A'
      }, {
        id: '2',
        correctness: undefined,
        content_html: 'B'
      }],
      type: 'student',
      answer_id: '',
      correct_answer_id: '',
      feedback_html: '',
      correct_answer_feedback_html: '',
      answered_count: 0,
      show_all_feedback: false,
      onChangeAnswer: jest.fn(),
      hideAnswers: false,
      hasCorrectAnswer: false,
      onChangeAttempt: jest.fn(),
      choicesEnabled: false,
    };
  });

  it('matches tutor teacher-preview snapshot', () => {
    props.answers.forEach((answer, i) => answer.content_html = answerContent[i]);
    const tree = renderer.create(
      <AnswersTable {...props} type="teacher-preview" onKeyPress={() => {}} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correct answer feedback', () => {
    const tree = renderer.create(
      <AnswersTable {...props}
        answer_id="1"
        correct_answer_id="1"
        correct_answer_feedback_html="Feedback"
        hasCorrectAnswer={true}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders incorrect answer feedback', () => {
    const tree = renderer.create(
      <AnswersTable {...props}
        answer_id="1"
        correct_answer_id="2"
        incorrectAnswerId="1"
        feedback_html="Feedback"
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('hides answers', () => {
    const tree = renderer.create(
      <AnswersTable {...props} hideAnswers={true} />
    ).toJSON();
    expect(tree).toBeNull();
  });

  it('generates an id if missing', () => {
    const tree = renderer.create(
      <AnswersTable {...props} questionId='' />
    );
    expect(tree.root.findAllByProps({ qid: 'auto-0' }).length).toBe(2);
  });
});
