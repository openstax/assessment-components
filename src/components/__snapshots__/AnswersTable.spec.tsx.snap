// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswersTable matches tutor teacher-preview snapshot 1`] = `
<div
  aria-label="Answer choices"
  className="answers-table"
  role="radiogroup"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
    >
      <div
        className="correct-incorrect"
      />
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
        onKeyPress={[Function]}
      >
        <span
          aria-label="Choice A:"
          className="answer-letter-wrapper"
          data-answer-choice="A"
          data-test-id="answer-choice-A"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
               
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ol>
  <li>matter is moving at speeds of less than roughly 1% the speed of light,</li>
  <li>objects are too small to be seen with the naked eye, and</li>
  <li>there is the involvement of only a weak gravitational field</li>
</ol>
",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
    >
      <div
        className="correct-incorrect"
      />
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
        onKeyPress={[Function]}
      >
        <span
          aria-label="Choice B:"
          className="answer-letter-wrapper"
          data-answer-choice="B"
          data-test-id="answer-choice-B"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
               
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<ol>
  <li>matter is moving at speeds greater than roughly 1% the speed of light,</li>
  <li>objects are large enough to be seen with the naked eye, and</li>
  <li>there is the involvement of a strong gravitational field</li>
</ol>
",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;

exports[`AnswersTable renders all feedback 1`] = `
<div
  aria-label="Answer choices"
  className="answers-table"
  role="radiogroup"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-selected answer-incorrect"
    >
      <input
        aria-details="feedback-1-0"
        checked={true}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          aria-label="Selected Choice A:"
          className="answer-letter-wrapper"
          data-answer-choice="A"
          data-test-id="answer-choice-A"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi faCGeG"
          >
            <span>
              Incorrect Answer
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <aside
    className="question-feedback bottom"
    id="feedback-1-0"
  >
    <div
      aria-label="Answer Feedback"
      className="arrow"
    />
    <aside>
      <div
        className="question-feedback-content has-html"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Answer level feedback",
          }
        }
      />
    </aside>
  </aside>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-correct"
    >
      <input
        aria-details="feedback-1-1"
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          aria-label="Choice B:"
          className="answer-letter-wrapper"
          data-answer-choice="B"
          data-test-id="answer-choice-B"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
              Correct Answer
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <aside
    className="question-feedback bottom"
    id="feedback-1-1"
  >
    <div
      aria-label="Answer Feedback"
      className="arrow"
    />
    <aside>
      <div
        className="question-feedback-content has-html"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Answer level feedback",
          }
        }
      />
    </aside>
  </aside>
</div>
`;

exports[`AnswersTable renders correct answer feedback 1`] = `
<div
  aria-label="Answer choices"
  className="answers-table"
  role="radiogroup"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-selected answer-correct"
    >
      <input
        aria-details="feedback-1-0"
        checked={true}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          aria-label="Selected Choice A:"
          className="answer-letter-wrapper"
          data-answer-choice="A"
          data-test-id="answer-choice-A"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
              Correct Answer
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <aside
    className="question-feedback bottom"
    id="feedback-1-0"
  >
    <div
      aria-label="Answer Feedback"
      className="arrow"
    />
    <aside>
      <div
        className="question-feedback-content has-html"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Feedback",
          }
        }
      />
    </aside>
  </aside>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          aria-label="Choice B:"
          className="answer-letter-wrapper"
          data-answer-choice="B"
          data-test-id="answer-choice-B"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
               
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;

exports[`AnswersTable renders incorrect answer feedback 1`] = `
<div
  aria-label="Answer choices"
  className="answers-table"
  role="radiogroup"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-selected answer-incorrect"
    >
      <input
        aria-details="feedback-1-0"
        checked={true}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          aria-label="Selected Choice A:"
          className="answer-letter-wrapper"
          data-answer-choice="A"
          data-test-id="answer-choice-A"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi faCGeG"
          >
            <span>
              Incorrect Answer
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <aside
    className="question-feedback bottom"
    id="feedback-1-0"
  >
    <div
      aria-label="Answer Feedback"
      className="arrow"
    />
    <aside>
      <div
        className="question-feedback-content has-html"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Feedback",
          }
        }
      />
    </aside>
  </aside>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled answer-correct"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          aria-label="Choice B:"
          className="answer-letter-wrapper"
          data-answer-choice="B"
          data-test-id="answer-choice-B"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
              Correct Answer
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;

exports[`AnswersTable renders instructions 1`] = `
<div
  aria-label="Answer choices"
  className="answers-table"
  role="radiogroup"
>
  <b>
    Instructions
  </b>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          aria-label="Choice A:"
          className="answer-letter-wrapper"
          data-answer-choice="A"
          data-test-id="answer-choice-A"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
               
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          aria-label="Choice B:"
          className="answer-letter-wrapper"
          data-answer-choice="B"
          data-test-id="answer-choice-B"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
               
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;

exports[`AnswersTable sorts by given ID order 1`] = `
<div
  aria-label="Answer choices"
  className="answers-table"
  role="radiogroup"
>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-0"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-0"
      >
        <span
          aria-label="Choice A:"
          className="answer-letter-wrapper"
          data-answer-choice="A"
          data-test-id="answer-choice-A"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
               
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "False",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
  <div
    className="openstax-answer"
  >
    <section
      className="answers-answer disabled"
    >
      <input
        checked={false}
        className="answer-input-box"
        disabled={true}
        id="1-option-1"
        name="1-options"
        onChange={[Function]}
        type="radio"
      />
      <label
        className="answer-label"
        htmlFor="1-option-1"
      >
        <span
          aria-label="Choice B:"
          className="answer-letter-wrapper"
          data-answer-choice="B"
          data-test-id="answer-choice-B"
        />
        <div
          aria-atomic="true"
          aria-live="polite"
          className="answer-answer"
          role="status"
        >
          <div
            className="sc-jSMfEi ecLwSd"
          >
            <span>
               
            </span>
          </div>
          <span
            className="answer-content"
            dangerouslySetInnerHTML={
              Object {
                "__html": "True",
              }
            }
          />
        </div>
      </label>
    </section>
  </div>
</div>
`;
